File: mod1lab3/somavetor.c
/* Multiplicador de matrizes quadradas
   Computação Concorrente
   Módulo 1 - Laboratório 2
   Pedro Henrique da Silva Bernardino  */
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include "timer.h"

float *vetor; //vetor do programa
float *result; //matriz de saída
int dim; //dimensão das matrizes quadradas
int nthreads; // número de threads

//funcao executada pelas threads
void *MultiplicaLinhaMatriz (void *linhaInicio) {

  int inicio = *((int *)linhaInicio);//linha da matriz em que a thread vai começar a operar
  for (int i = inicio; i < dim; i+= nthreads)
  	//efetua a multiplicação de uma linha da mat1 pela mat2;
  	for(int j = 0; j < dim; j++)
  		for(int k = 0; k < dim; k++)
  			*(result+(dim*i)+j) += *(mat1+(dim*i)+k) * *(mat2+(dim*k)+j);


  pthread_exit(NULL);
}


int main(int argc, char* argv[])
{
	int *linhas;
	pthread_t *tid;
	double inicio, parcial, fim, elapsed;

	//recebe os argumentos das linhas de comando
	if (argc<3)
	{
		printf("Digite a dimensão das matrizes e o número de threads ao executar o arquivo.\n");
		return 1;
	}
	dim = atoi(argv[1]);
	nthreads = atoi(argv[2]);

	//cria as matrizes usadas no programa
	GET_TIME(inicio);
	parcial = inicio;

	tid = (pthread_t*) malloc (sizeof(pthread_t) * nthreads);
	if (tid == NULL) {
      printf("--ERRO: malloc()\n"); exit(-1);
    }

	mat1 = (float *) malloc (sizeof(float) * dim * dim);
	if (mat1 == NULL) {
      printf("--ERRO: malloc()\n"); exit(-1);
    }

	mat2 = (float *) malloc (sizeof(float) * dim * dim);
	if (mat2 == NULL) {
      printf("--ERRO: malloc()\n"); exit(-1);
    }

	result = (float *) calloc (dim*dim, sizeof(float));
	if (result == NULL) {
      printf("--ERRO: malloc()\n"); exit(-1);
    }

	linhas = (int*) malloc(sizeof(int)*nthreads);
	if (linhas == NULL) {
      printf("--ERRO: malloc()\n"); exit(-1);
    }

	// gera valores para as matrizes de entrada
	for (int i = 0; i < dim; ++i)
	{
		for (int j = 0; j < dim; ++j)
		{
			*(mat1+(dim*i)+j) = 2*i+4*j;
			*(mat2+(dim*i)+j) = 3*j-2*i;
		}
	}

	GET_TIME(fim);
	elapsed = fim - parcial;
    printf("O tempo de inicialização das estruturas de dados foi de %lf segundos. \n", elapsed);
    parcial = fim;

	//cria as threads
	for (int thread = 0; thread < nthreads; ++thread)
	{
		*(linhas+thread) = thread;  
		if (pthread_create(tid+thread, NULL, MultiplicaLinhaMatriz, (void*)(linhas+thread)))
		{
      		printf("--ERRO: pthread_create()\n"); exit(-1);
    	}
	}

	//--espera todas as threads terminarem
	for (int thread=0; thread<nthreads; thread++) 
	{
    	if (pthread_join(*(tid+thread), NULL)) 
    	{
        	printf("--ERRO: pthread_join() \n");
        	exit(-1); 
        }
    } 

    GET_TIME(fim);
	elapsed = fim - parcial;
    printf("O tempo de cálculo da multiplicação pelas threads foi de %lf segundos. \n", elapsed);
    parcial = fim;

  	//printa o resultado da multiplicação - desativado por ser impossível checar os tempos com o programa imprimindo uma matriz 500x500 no processo.
  	
	for (int i = 0; i < dim; ++i)
	{
		printf("|");
		for (int j = 0; j < dim; ++j)
		{
			printf("%.2f  ", *(mat1+(dim*i)+j));
		}
		printf("|");
		printf("\t");
		printf("|");
		for (int j = 0; j < dim; ++j)
		{
			printf("%.2f  ", *(mat2+(dim*i)+j));
		}
		printf("|");
		printf("\t");
		printf("|");
		for (int j = 0; j < dim; ++j)
		{
			printf("%.2f  ", *(result+(dim*i)+j));
		}
		printf("|");
		printf("\n");
		printf("|");
	}
	

	GET_TIME(fim);
	
	elapsed = fim - parcial;
    printf("O tempo de exibição dos resultados foi de %lf segundos. \n", elapsed);
    
    elapsed = fim - inicio;
    printf("O tempo total de execução do programa foi de %lf segundos. \n", elapsed);

	return 0;
}

File: mod2lab6/somavetor.java
/* Disciplina: Computacao Concorrente */
/* Aluno: Pedro Henrique Bernardino */
/* Módulo 2 - Laboratório 6 */
/* Soma de vetores usando computação concorrente */

import java.util.Random;




//classe da Thread
class SomaVetor extends Thread {
    private int id;
    int N; // número de Threads
    int TAM; // Tamanho dos vetores;
    int[] vet1; //primeiro vetor da soma
    int[] vet2; //segundo vetor da soma
    int[] result; //vetor resultado da soma

    //--construtor - Thread recebe seu id
    public SomaVetor(int id, int N, int TAM, int[] vet1, int[] vet2, int[] result) {
        this.id = id;
        this.N = N;
        this.TAM = TAM;
        this.vet1 = vet1;
        this.vet2 = vet2;
        this.result  = result;
    }

    //--metodo executado pela thread
    public void run() {
        for(int i = id; i< TAM; i+= N)
        {
            result[i] = vet1[i] + vet2[i];
        }
    }
}

//classe principal do programa, que gerencia as threads
class Main
{


    public static void main(String[] args) {

        int N = 4; // número de Threads
        int TAM = 10; // Tamanho dos vetores;
        int[] vet1 = new int[TAM]; //primeiro vetor da soma
        int[] vet2 = new int[TAM]; //segundo vetor da soma
        int[] result = new int[TAM]; //vetor resultado da soma

        Thread[] threads = new Thread[N];

        //cria as threads
        for (int i=0; i<threads.length; i++) {
            threads[i] = new SomaVetor(i, N, TAM, vet1, vet2, result);
        }

        //inicializa os vetores
        for(int i=0; i<TAM; i++) {
            vet1[i] = new Random().nextInt(100);
            vet2[i] = new Random().nextInt(100);
        }

        //inicia as threads
        for (int i=0; i<threads.length; i++) {
            threads[i].start();
        }

        //espera as threads encerrarem
        for (int i=0; i<threads.length; i++) {
            try { threads[i].join(); }
            catch (InterruptedException e) { return; }
        }


        //exibição dos resultados
        for(int i=0; i<TAM; i++) {
            System.out.print(vet1[i] + " ");
        }
        System.out.println("");
        for(int i=0; i<TAM; i++) {
            System.out.print(vet2[i] + " ");
        }
        System.out.println("");
        for(int i=0; i<TAM; i++) {
            System.out.print(result[i] + " ");
        }
    }
}